hoodie-client/src/main/java/com/uber/hoodie/HoodieWriteClient.java

```diff
@@ -414,7 +414,11 @@ private void saveWorkloadProfileMetadataToInflight(WorkloadProfile profile,
       final boolean isUpsert) {
 
     // Cache the tagged records, so we don't end up computing both
-    preppedRecords.persist(StorageLevel.MEMORY_AND_DISK_SER());
+    if (preppedRecords.getStorageLevel() == StorageLevel.NONE()) {
```

Comment:
The code change looks good but I have general concern about the API contract here. 

HoodieWriteClient is one of the main user-facing class. There are a bunch of APIs (insert/upsert and their bulk versions) which takes in non-prepped records and another set with prepped records. In all cases, the APIs get a JavaRDD as input.  

We need to give a consistent contract (like persist side-effect from operations perspective). In the case of non-prepped records, we just apply transformations. 

To make them consistent, I see 2 options 

1. when Hudi persists caller's prepped RDD, Hudi will unpersist() before it returns the control back to the caller (preferred choice) or
2. Make an explicit ownership contract where Hudi never touches client provided RDD and it is the responsibility of the caller to cache/persist.

Reply:
@bvaradar Good point on making a consistent contract in HoodieWriteClient.  As currently we don't do unpersist of `preppedRecords`,  so the effect of this PR is to support a use case where client has persisted `preppedRecords` without other additional changes.

I'm in favor of suggestion #1 of unpersist if Hudi has persisted the record.  However since this involves a contract change, can we consider to make this change in a separate issue and have the new contract reviewed and tested by all clients?

Reply:
@jianxu @bvaradar For option 1, we should look at the performance overhead of doing a persist() and unpersist() if eventually the client wants to anyways persist.

Reply:
@n3nash : In ingestion use-cases, I was expecting that writing to Hoodie (sink) will be the last part most of the time. But in any case in practical sense, Isn't prepping records (outside Hoodie) usually a non-trivial work and it is a good idea to persist as soon as it is ready to prevent rework in case of failures ? I am fine with (2) too as it gives flexibility and ownership to clients.  

@jianxu : I am ok with doing the change in a separate PR as this PR is more of a bug fix. Can you open an issue to do the cleanup and add a TODO  here. 

Balaji.V

Reply:
Good points @bvaradar @n3nash. And https://github.com/uber/hudi/issues/362 is to track the consistent contract.